pkg/adapter/awssnssource/adapter.go

```diff
@@ -18,216 +18,155 @@ package awssnssource
 
 import (
 	"context"
-	"encoding/json"
 	"fmt"
-	"io/ioutil"
 	"net/http"
 	"time"
 
 	"go.uber.org/zap"
 
 	cloudevents "github.com/cloudevents/sdk-go/v2"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/arn"
-	"github.com/aws/aws-sdk-go/aws/session"
-	"github.com/aws/aws-sdk-go/service/sns"
-	"github.com/aws/aws-sdk-go/service/sns/snsiface"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
 
 	pkgadapter "knative.dev/eventing/pkg/adapter/v2"
 	"knative.dev/pkg/apis"
+	k8sclient "knative.dev/pkg/client/injection/kube/client"
 	"knative.dev/pkg/injection"
 	"knative.dev/pkg/logging"
 
+	"github.com/triggermesh/aws-event-sources/pkg/adapter/awssnssource/handler"
 	"github.com/triggermesh/aws-event-sources/pkg/adapter/awssnssource/status"
-	"github.com/triggermesh/aws-event-sources/pkg/adapter/common"
 	"github.com/triggermesh/aws-event-sources/pkg/adapter/common/env"
+	"github.com/triggermesh/aws-event-sources/pkg/adapter/common/router"
 	"github.com/triggermesh/aws-event-sources/pkg/apis/sources/v1alpha1"
+	"github.com/triggermesh/aws-event-sources/pkg/client/generated/injection/client"
+	snsclient "github.com/triggermesh/aws-event-sources/pkg/client/sns"
+	"github.com/triggermesh/aws-event-sources/pkg/routing"
 )
 
-// envConfig is a set parameters sourced from the environment for the source's
-// adapter.
-type envConfig struct {
-	env.Config
-
-	ARN string `required:"true"`
-}
-
 // adapter implements the source's adapter.
 type adapter struct {
 	logger *zap.SugaredLogger
 
-	snsClient snsiface.SNSAPI
-	ceClient  cloudevents.Client
-
-	arn arn.ARN
+	ceClient cloudevents.Client
+	snsCg    snsclient.ClientGetter
 
 	// fields accessed during object reconciliation
+	router        *router.Router
 	statusPatcher *status.Patcher
 }
```

Comment:
The adapter now includes:
* a _SNS client getter_ that returns SNS clients based on the AWS region indicated in the ARN of the reconciled source
* a _request router_ that routes requests originating from SNS to their corresponding handler